src/daft-scan/src/lib.rs

```diff
@@ -1,93 +1,318 @@
+    pub columns: Option<Arc<Vec<String>>>,
+    pub limit: Option<usize>,
+    pub metadata: Option<TableMetadata>,
+    pub statistics: Option<TableStatistics>,
+}
+
+impl ScanTaskBatch {
+    pub fn new(
+        sources: Vec<DataFileSource>,
+        file_format_config: Arc<FileFormatConfig>,
+        schema: SchemaRef,
+        storage_config: Arc<StorageConfig>,
+        columns: Option<Arc<Vec<String>>>,
+        limit: Option<usize>,
+    ) -> Self {
+        assert!(!sources.is_empty());
+        let (length, statistics) = sources
+            .iter()
+            .map(|s| {
+                (
+                    s.get_metadata().map(|m| m.length),
+                    s.get_statistics().cloned(),
+                )
+            })
+            .reduce(|(acc_len, acc_stats), (curr_len, curr_stats)| {
+                (
+                    acc_len.and_then(|acc_len| curr_len.map(|curr_len| acc_len + curr_len)),
+                    acc_stats.and_then(|acc_stats| {
+                        curr_stats.map(|curr_stats| acc_stats.union(&curr_stats).unwrap())
+                    }),
+                )
+            })
+            .unwrap();
+        let metadata = length.map(|l| TableMetadata { length: l });
+        Self {
+            sources,
+            file_format_config,
+            schema,
+            storage_config,
+            columns,
+            limit,
+            metadata,
+            statistics,
+        }
+    }
+
+    pub fn num_rows(&self) -> Option<usize> {
+        self.metadata.as_ref().map(|m| m.length)
+    }
+
+    pub fn size_bytes(&self) -> Option<usize> {
+        self.statistics.as_ref().and_then(|s| {
+            self.num_rows()
+                .and_then(|num_rows| Some(num_rows * s.estimate_row_size().ok()?))
+        })
+    }
+}
+
+impl From<Vec<ScanTask>> for ScanTaskBatch {
+    fn from(value: Vec<ScanTask>) -> Self {
+        if value.is_empty() {
+            panic!("Must have at least one ScanTask to create a ScanTaskBatch.");
+        }
+        let mut scan_task_iter = value.into_iter();
+        let first_scan_task = scan_task_iter.next().unwrap();
+        let first_scan_task_source = first_scan_task.source;
+        let sources = vec![first_scan_task_source]
+            .into_iter()
+            .chain(scan_task_iter.map(|t| t.source))
+            .collect::<Vec<_>>();
+        Self::new(
+            sources,
+            first_scan_task.file_format_config,
+            first_scan_task.schema,
+            first_scan_task.storage_config,
+            first_scan_task.columns,
+            first_scan_task.limit,
+        )
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
 pub struct PartitionField {
     field: Field,
     source_field: Option<Field>,
     transform: Option<Expr>,
 }
 
-pub trait ScanOperator: Send + Display {
+pub trait ScanOperator: Send + Sync + Display + Debug + DynHash {
```

Comment:
Do we still need `DynHash`? I was thinking for the logical plan equality / hashing we could leverage the pointer address of the `Box<ScanOperator>` since just care if it is the same instance.

Reply:
See: https://github.com/Eventual-Inc/Daft/pull/1562